Objective:                                                              
1. To build a system that recommends new music to users in this community from the available information by calculating support, confidence, and lift.                                                 
2. To look for artists (or groups
of artists) who have support larger than 0.01 (1%) and who give confidence to
another artist that is larger than 0.50 (50%) and also lift that is larger than 5            

In [1]:
import pandas as pd
import numpy as np
from mlxtend.frequent_patterns import apriori,association_rules
pd.options.display.max_columns = 25
pd.options.display.max_rows = None

In [2]:
lastfm = pd.read_csv("https://www.biz.uiowa.edu/faculty/jledolter/DataMining/lastfm.csv")
lastfm_copy = lastfm.copy()

In [3]:
lastfm.head(10)

user                   artist sex  country
0     1    red hot chili peppers   f  Germany
1     1  the black dahlia murder   f  Germany
2     1                goldfrapp   f  Germany
3     1         dropkick murphys   f  Germany
4     1                 le tigre   f  Germany
5     1               schandmaul   f  Germany
6     1                    edguy   f  Germany
7     1             jack johnson   f  Germany
8     1                eluveitie   f  Germany
9     1              the killers   f  Germany

In [4]:
#to check if null are there
lastfm.isnull().sum()

user       0
artist     0
sex        0
country    0
dtype: int64

In [5]:
#to get no: of rows and columns
lastfm.shape

(289955, 4)

In [6]:
#to get the no:of duplicated rows
lastfm.duplicated().sum()

2

In [7]:
#deleted the duplicated rows
lastfm = lastfm.drop_duplicates()

In [8]:
#check the shape of dataset after deletion of duplicate rows
lastfm.shape

(289953, 4)

In [9]:
#lastfm['cnt'] = 1

In [10]:
#using group by to group the artist column  with respect to user .Using count to get the count as 1 and make it horizontal
lastfm_1 = lastfm.groupby(['user','artist'])['user'].count().unstack()

In [11]:
lastfm_1.head(10)

artist  ...and you will know us by the trail of dead  2pac  3 doors down  \
user                                                                       
1                                                NaN   NaN           NaN   
3                                                NaN   NaN           NaN   
4                                                NaN   NaN           NaN   
5                                                NaN   NaN           NaN   
6                                                NaN   NaN           NaN   
7                                                NaN   NaN           NaN   
9                                                NaN   NaN           NaN   
12                                               NaN   NaN           NaN   
13                                               NaN   NaN           NaN   
14                                               NaN   NaN           NaN   

artist  30 seconds to mars  311  36 crazyfists  44  50 cent  65daysofstatic  \
user                                                                          
1                      NaN  NaN            NaN NaN      NaN             NaN   
3                      NaN  NaN            NaN NaN      NaN             NaN   
4                      NaN  NaN            NaN NaN      NaN             NaN   
5                      NaN  NaN            NaN NaN      NaN             NaN   
6                      NaN  NaN            NaN NaN      NaN             NaN   
7                      NaN  NaN            NaN NaN      NaN             NaN   
9                      NaN  NaN            NaN NaN      NaN             NaN   
12                     NaN  NaN            NaN NaN      NaN             NaN   
13                     NaN  NaN            NaN NaN      NaN             NaN   
14                     NaN  NaN            NaN NaN      NaN             NaN   

artist  [unknown]  a day to remember  a fine frenzy  ...  volbeat  weezer  \
user                                                 ...                    
1             NaN                NaN            NaN  ...      NaN     NaN   
3             NaN                NaN            NaN  ...      NaN     NaN   
4             NaN                NaN            NaN  ...      NaN     NaN   
5             NaN                NaN            NaN  ...      1.0     NaN   
6             NaN                NaN            NaN  ...      NaN     NaN   
7             NaN                NaN            NaN  ...      NaN     NaN   
9             NaN                NaN            NaN  ...      NaN     NaN   
12            NaN                NaN            NaN  ...      NaN     NaN   
13            NaN                NaN            NaN  ...      NaN     NaN   
14            NaN                NaN            NaN  ...      NaN     NaN   

artist  wilco  within temptation  wolfgang amadeus mozart  wu-tang clan  \
user                                                                      
1         NaN                NaN                      NaN           NaN   
3         NaN                NaN                      NaN           NaN   
4         NaN                NaN                      NaN           NaN   
5         NaN                NaN                      NaN           NaN   
6         NaN                NaN                      NaN           NaN   
7         NaN                NaN                      NaN           NaN   
9         NaN                NaN                      NaN           NaN   
12        NaN                NaN                      NaN           NaN   
13        NaN                NaN                      NaN           NaN   
14        NaN                NaN                      NaN           NaN   

artist  yann tiersen  yeah yeah yeahs  yellowcard  yo la tengo  zero 7  \
user                                                                     
1                NaN              NaN         NaN          NaN     NaN   
3                NaN              NaN         NaN          NaN     NaN   
4               

In [12]:
#defining a function to convert all the NaN values to 0
def make_boolean(x):
    if x>0:
        return 1
    else:
        return 0

In [13]:
#using applymap function to use the created function in the dataset
lastfm_final = lastfm_1.applymap(make_boolean)

In [14]:
#matrix is formed and checked the shape
lastfm_final.shape

(15000, 1004)

In [15]:
lastfm_final.head(10)

artist  ...and you will know us by the trail of dead  2pac  3 doors down  \
user                                                                       
1                                                  0     0             0   
3                                                  0     0             0   
4                                                  0     0             0   
5                                                  0     0             0   
6                                                  0     0             0   
7                                                  0     0             0   
9                                                  0     0             0   
12                                                 0     0             0   
13                                                 0     0             0   
14                                                 0     0             0   

artist  30 seconds to mars  311  36 crazyfists  44  50 cent  65daysofstatic  \
user                                                                          
1                        0    0              0   0        0               0   
3                        0    0              0   0        0               0   
4                        0    0              0   0        0               0   
5                        0    0              0   0        0               0   
6                        0    0              0   0        0               0   
7                        0    0              0   0        0               0   
9                        0    0              0   0        0               0   
12                       0    0              0   0        0               0   
13                       0    0              0   0        0               0   
14                       0    0              0   0        0               0   

artist  [unknown]  a day to remember  a fine frenzy  ...  volbeat  weezer  \
user                                                 ...                    
1               0                  0              0  ...        0       0   
3               0                  0              0  ...        0       0   
4               0                  0              0  ...        0       0   
5               0                  0              0  ...        1       0   
6               0                  0              0  ...        0       0   
7               0                  0              0  ...        0       0   
9               0                  0              0  ...        0       0   
12              0                  0              0  ...        0       0   
13              0                  0              0  ...        0       0   
14              0                  0              0  ...        0       0   

artist  wilco  within temptation  wolfgang amadeus mozart  wu-tang clan  \
user                                                                      
1           0                  0                        0             0   
3           0                  0                        0             0   
4           0                  0                        0             0   
5           0                  0                        0             0   
6           0                  0                        0             0   
7           0                  0                        0             0   
9           0                  0                        0             0   
12          0                  0                        0             0   
13          0                  0                        0             0   
14          0                  0                        0             0   

artist  yann tiersen  yeah yeah yeahs  yellowcard  yo la tengo  zero 7  \
user                                                                     
1                  0                0           0            0       0   
3                  0                0           0            0       0   
4               

In [16]:
#minimum support set is 0.01.Using apriori function to filter rows according to support and item set created according to its frequency of occurance
ap = apriori(lastfm_final, min_support = .01, use_colnames = True)

In [17]:
#obtained support for the itemsets
ap.head()

support              itemsets
0  0.022733                (2pac)
1  0.030933        (3 doors down)
2  0.032800  (30 seconds to mars)
3  0.021800             (50 cent)
4  0.013667      (65daysofstatic)

In [18]:
#minimum confidence set as 0.5.Here metric is set as confidence
rules = association_rules(ap, metric = "confidence",min_threshold = .5)

In [19]:
rules.head()

antecedents  consequents  antecedent support  consequent support  \
0                 (beck)  (radiohead)            0.057467            0.180267   
1                 (blur)  (radiohead)            0.033533            0.180267   
2  (broken social scene)  (radiohead)            0.027533            0.180267   
3                (keane)   (coldplay)            0.034933            0.158533   
4          (snow patrol)   (coldplay)            0.050400            0.158533   

    support  confidence      lift  leverage  conviction  
0  0.029267    0.509281  2.825152  0.018907    1.670473  
1  0.017533    0.522863  2.900496  0.011488    1.718024  
2  0.015067    0.547215  3.035589  0.010103    1.810427  
3  0.022267    0.637405  4.020634  0.016729    2.320676  
4  0.026467    0.525132  3.312441  0.018477    1.772002

In [20]:
#now filtered with lift greater than 5 and sorting according to decresing oredr of confidence
rules[rules["lift"]>5].sort_values(by='confidence',ascending=False)
#here obtained 6 associations with support > 0.01 and confidence > .50 and lift > 5


antecedents     consequents  antecedent support  \
38  (led zeppelin, the doors)    (pink floyd)            0.017867   
12       (the pussycat dolls)       (rihanna)            0.018000   
8                      (t.i.)    (kanye west)            0.018333   
39    (the doors, pink floyd)  (led zeppelin)            0.019800   
10           (sonata arctica)     (nightwish)            0.026400   
7              (judas priest)   (iron maiden)            0.026667   

    consequent support   support  confidence       lift  leverage  conviction  
38            0.104933  0.010667    0.597015   5.689469  0.008792    2.221091  
12            0.043067  0.010400    0.577778  13.415893  0.009625    2.266421  
8             0.064067  0.010400    0.567273   8.854413  0.009225    2.162871  
39            0.079200  0.010667    0.538721   6.802027  0.009099    1.996187  
10            0.061933  0.013467    0.510101   8.236292  0.011832    1.914816  
7             0.059267  0.013533    0.507500   8.562992  0.011953    1.910118

**SUMMARY**

--> the pussycat dolls and rihanna gives an individual support of 1.8% and 4.3% 
respectively.

--> probability of buying  pussycat dolls and rihanna 0.0104. ie, 1.04% and

--> confidence = probability of (buying rihanna given the pussycat dolls is bought)=0.577 . ie, 57.77% with a highest lift of 13.41.

--> This association is the best among the available.